File: frontend/src/features/connectivity/browserPolicy.ts
/**
 * Browser support policy evaluation (pre-M6 minimum-version enforcement).
 *
 * A pure function maps a detected browser + the backend-provided policy to a typed decision.
 * Version detection is a compatibility/supportability gate, NOT a security boundary; it never
 * influences liveness/PASS/portrait/fraud/callback security.
 */

import type { BrowserFamily, BrowserInfo } from './browserDetection'

export interface BrowserPolicyEntry {
  minimum_major: number
  enabled: boolean
}

export interface BrowserSupportPolicy {
  policy_version: string
  browsers: Record<Exclude<BrowserFamily, 'unknown'>, BrowserPolicyEntry>
}

export type PolicyDecisionState =
  'SUPPORTED' | 'VERSION_TOO_OLD' | 'UNSUPPORTED_BROWSER' | 'VERSION_UNKNOWN' | 'POLICY_UNAVAILABLE'

export interface PolicyDecision {
  state: PolicyDecisionState
  /** Internal metadata — never shown verbatim to customers. */
  browserFamily: BrowserFamily
  detectedMajorVersion: number | null
  requiredMajorVersion: number | null
  policyVersion: string | null
}

/**
 * Evaluate a detected browser against the backend policy.
 *
 * - policy is null/absent           -> POLICY_UNAVAILABLE (fail closed)
 * - family unknown                  -> UNSUPPORTED_BROWSER
 * - family recognized but disabled  -> UNSUPPORTED_BROWSER
 * - recognized family + null version -> VERSION_UNKNOWN (never invent a version)
 * - version below minimum           -> VERSION_TOO_OLD
 * - otherwise                       -> SUPPORTED
 */
export function evaluateBrowserPolicy(
  browser: BrowserInfo,
  policy: BrowserSupportPolicy | null,
): PolicyDecision {
  const base = {
    browserFamily: browser.family,
    detectedMajorVersion: browser.majorVersion,
    requiredMajorVersion: null,
    policyVersion: policy?.policy_version ?? null,
  }

  if (policy === null) {
    return { ...base, state: 'POLICY_UNAVAILABLE' }
  }
  if (browser.family === 'unknown') {
    return { ...base, state: 'UNSUPPORTED_BROWSER' }
  }

  const entry = policy.browsers[browser.family]
  if (!entry || !entry.enabled) {
    return { ...base, state: 'UNSUPPORTED_BROWSER' }
  }

  const withRequired = { ...base, requiredMajorVersion: entry.minimum_major }
  if (browser.majorVersion === null) {
    return { ...withRequired, state: 'VERSION_UNKNOWN' }
  }
  if (browser.majorVersion < entry.minimum_major) {
    return { ...withRequired, state: 'VERSION_TOO_OLD' }
  }
  return { ...withRequired, state: 'SUPPORTED' }
}

File: backend/app/core/browser_policy.py
"""Backend-owned browser support policy (pre-M6).

The minimum supported browser versions are operational configuration owned by the backend, not
hardcoded in React components and not environment-secret values. The committed
``config/browser-support.json`` holds PROVISIONAL compatibility baselines (not certification
claims). Later the file can be supplied by a Kubernetes ConfigMap without code changes.

Security note: this policy is a compatibility/supportability gate, NOT a security boundary.
Browser/version detection can be spoofed and must never influence liveness, PASS authorization,
portrait authorization, fraud decisions, or callback security.

Reload semantics: the policy is loaded once at application startup (see ``app/factory.py``).
Changing ``config/browser-support.json`` requires a backend restart / pod rollout. This is the
documented, predictable behaviour for pre-M6; no ad-hoc hot reload is implemented.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, StrictBool, ValidationError

BrowserFamily = Literal[
    "chrome",
    "edge",
    "firefox",
    "safari",
    "ios_safari",
    "android_chrome",
]


class BrowserPolicyConfigError(Exception):
    """Raised when the browser-support policy file is missing, unparseable or invalid."""


class BrowserPolicyEntry(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # Strict types: a malformed entry (string number, "yes"/"1" booleans) is rejected, never
    # silently coerced into a plausible-looking policy.
    minimum_major: int = Field(
        gt=0, strict=True, description="minimum supported major version (>=1)"
    )
    enabled: StrictBool


class BrowserPolicies(BaseModel):
    """Exactly the six known browser families; unknown keys are rejected."""

    model_config = ConfigDict(extra="forbid")

    chrome: BrowserPolicyEntry
    edge: BrowserPolicyEntry
    firefox: BrowserPolicyEntry
    safari: BrowserPolicyEntry
    ios_safari: BrowserPolicyEntry
    android_chrome: BrowserPolicyEntry


class BrowserSupportPolicy(BaseModel):
    """Typed browser support policy, safe to expose via /api/v1/info.

    Only non-sensitive policy data is present: no filesystem paths, secrets or internal
    deployment information.
    """

    model_config = ConfigDict(extra="forbid")

    policy_version: str = Field(min_length=1)
    browsers: BrowserPolicies


def load_browser_support_policy(path: str) -> BrowserSupportPolicy:
    """Load and validate the browser-support policy from a JSON file.

    Raises ``BrowserPolicyConfigError`` on missing/unparseable/invalid file so callers can fail
    closed. A malformed policy is never silently accepted.
    """
    config_path = Path(path)
    try:
        raw = config_path.read_bytes()
    except OSError as exc:
        raise BrowserPolicyConfigError(
            f"cannot read browser-support policy at '{path}': {exc}"
        ) from exc
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise BrowserPolicyConfigError(f"browser-support policy JSON is invalid: {exc}") from exc
    if not isinstance(parsed, dict):
        raise BrowserPolicyConfigError("browser-support policy must be a JSON object")
    try:
        return BrowserSupportPolicy.model_validate(parsed)
    except ValidationError as exc:
        raise BrowserPolicyConfigError(f"browser-support policy is invalid: {exc}") from exc
